### PR TITLE
PLANNER-2313 Solver REST API draft

### PR DESCRIPTION
**Pull request by rsynek**
I am opening this draft PR to illustrate how the REST API might look like on one of the quickstarts. The next step is to generalize reusable parts of the code and generate the rest by Gizmo.

This PR will stay as a draft and will be updated/replaced by the generated endpoint.

### JIRA

https://issues.redhat.com/browse/PLANNER-2313

### CODE REVIEW

File: quarkus-school-timetabling/src/main/java/org/acme/schooltimetabling/persistence/TimeTableRepository.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+import javax.transaction.Transactional;
+
+import org.acme.schooltimetabling.domain.Lesson;
+import org.acme.schooltimetabling.domain.TimeTable;
+import org.optaplanner.SolutionRepository;
+
+import io.quarkus.panache.common.Sort;
+
+@ApplicationScoped
+public class TimeTableRepository implements SolutionRepository<TimeTable, Long> {
```
Comment (rsynek) [inline quarkus-school-timetabling/src/main/java/org/acme/schooltimetabling/persistence/TimeTableRepository.java:16]: This is what the user is expected to provide: a repository for the planning solution.

File: quarkus-school-timetabling/src/main/java/org/acme/schooltimetabling/rest/TimeTableResource.java [hunk 2]
```diff
@@ -16,97 +16,224 @@
...
-@Consumes(MediaType.APPLICATION_JSON)
 public class TimeTableResource {
 
-    public static final Long SINGLETON_TIME_TABLE_ID = 1L;
+    private static final String OPEN_API_PROBLEM_ID_DESC = "Never null, an ID for each planning problem. This must be unique.";
+
+    private static final String PARAM_PROBLEM_ID = "problemId";
+
+    // TODO: If the client calls only solve(), there is a memory leak.
+    private Map<Long, Throwable> errors = new ConcurrentHashMap<>();
```
Comment (rsynek) [inline quarkus-school-timetabling/src/main/java/org/acme/schooltimetabling/rest/TimeTableResource.java:55]: Here comes the interesting part: the `SolverManager.solve()` does not throw exceptions during solving; it passes them to a consumer provided by the user. This is a conflict between the "reactive" and "polling" approach.

The execution error is available via the `getError()` call and also any polling call (`getStatus()`, `getTimeTable()` and `stopSolving()`) will result with the `INTERNAL_SERVER_ERROR` response, until the error for given `planningId` is cleared by calling to `deleteError()`.

TLDR; is it very difficult not to notice the error.

Despite that, if someone is calling `solve()` in a loop without caring about the errors, there is a memory leak.

File: quarkus-school-timetabling/src/main/java/org/optaplanner/SolutionRepository.java [hunk 3]
```diff
@@ -0,0 +1,8 @@
+package org.optaplanner;
+
+public interface SolutionRepository<Solution, ProblemId> {
```
Comment (rsynek) [inline quarkus-school-timetabling/src/main/java/org/optaplanner/SolutionRepository.java:3]: This will end up in the runtime module of the new extension (with proper javadoc, of course).
